streams/core.py

```diff
@@ -474,6 +483,8 @@ def update(self, x, who=None):
         if len(L) == 1 and all(self.buffers):
             tup = tuple(buf.popleft() for buf in self.buffers)
             self.condition.notify_all()
+            if tup and hasattr(tup[0], '__stream_merge__'):
+                tup = tup[0].__stream_merge__(*tup[1:])
```

Comment:
In the future this would need some sort of dispatching mechanism to combine different kinds of objects.